File: libs/java/src/main/java/com/mbreissi/edgecommons/facades/DataFacade.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.mbreissi.edgecommons.facades;

import com.mbreissi.edgecommons.config.ConfigManager;
import com.mbreissi.edgecommons.messaging.Message;
import com.mbreissi.edgecommons.messaging.MessageBuilder;
import com.mbreissi.edgecommons.messaging.MessagingClient;
import com.mbreissi.edgecommons.messaging.Qos;
import com.mbreissi.edgecommons.uns.Uns;
import com.mbreissi.edgecommons.uns.UnsClass;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.nio.charset.StandardCharsets;
import java.time.Clock;
import java.time.Instant;
import java.util.Objects;

/**
 * The {@code data()} publish facade — the telemetry / signal data plane (DESIGN-class-facades §2.1,
 * D2/D5). It <b>constructs and validates the {@code SouthboundSignalUpdate} body</b>
 * ({@code device}/{@code signal}/{@code samples}) so an adapter never hand-builds it, applies the
 * body defaults, sanitizes the signal path into the UNS {@code data} channel, stamps the envelope
 * identity, and routes to the resolved {@link Channel}. It publishes through the <b>ordinary,
 * guarded</b> {@code messaging().publish(...)} — {@code data} is non-reserved, so it passes the
 * guard; the facade adds body-contract enforcement + defaults, <b>not</b> privilege (it is NOT a
 * {@code ReservedPublisher}).
 *
 * <p><b>Body ({@code header.name} = {@value #DATA_MESSAGE_NAME}, version {@value #DATA_MESSAGE_VERSION}):</b>
 * <pre>{@code
 *   { "device": {adapter, instance, endpoint}?,          // optional block
 *     "signal": { "id": <REQUIRED>, "name"?, "address"? },
 *     "samples": [ { "value": <REQUIRED>, "quality", "qualityRaw"?, "sourceTs"?, "serverTs" } ] }
 * }</pre>
 *
 * <p><b>Defaulting (DESIGN-class-facades §2.1, pinned by {@code uns-test-vectors/data.json}):</b>
 * <ol>
 *   <li>{@code quality} → {@link Quality#GOOD} when omitted on a sample that carries a value.</li>
 *   <li>{@code qualityRaw} → the synthetic marker {@value #QUALITY_UNSPECIFIED} when (and only
 *       when) the quality was defaulted; else the caller's value verbatim, else absent.</li>
 *   <li>{@code serverTs} → now (ISO-8601 UTC {@code …Z}, from the injected {@link Clock}) when
 *       omitted; {@code sourceTs} is <b>never</b> synthesized (absent when the source has none).</li>
 *   <li>the {@code samples} wrapper is enforced for the value-shorthand (a caller never emits a
 *       bare value).</li>
 *   <li>{@code signal.id} is the <b>only</b> hard reject — a publish with no stable id throws
 *       {@link IllegalArgumentException} at the call site.</li>
 * </ol>
 *
 * <p><b>Channel routing (DESIGN-class-facades §4, D1):</b> per-call {@link SignalUpdate.Builder#via}
 * override ▸ config {@code publish.channel} (instance ▸ global) ▸ {@link Channel#LOCAL}. A
 * {@code stream:<name>} route serializes the same envelope and appends it to
 * {@code getStreams().stream(name)} (partition key = {@code signal.id}, ts = {@code serverTs}); when
 * streaming is not configured it falls back to a LOCAL publish (readiness / no-streaming → local).
 * Northbound / stream transport failures are caught and logged — they must never flip local
 * readiness.
 *
 * <p><b>Library-internal:</b> obtain the bound instance via {@code gg.instance(id).data()} (or the
 * {@code main}-instance convenience {@code gg.getData()}); the public constructor exists only so the
 * per-instance handle in {@code com.mbreissi.edgecommons} can wire it.
 */
public final class DataFacade {

    private static final Logger LOGGER = LogManager.getLogger(DataFacade.class);

    /** The signal-update envelope header name ({@code docs/SOUTHBOUND.md} §2). */
    public static final String DATA_MESSAGE_NAME = "SouthboundSignalUpdate";
    /** The signal-update envelope header version. */
    public static final String DATA_MESSAGE_VERSION = "1.0";
    /** The {@code qualityRaw} marker written when {@code quality} was defaulted to {@code GOOD}. */
    public static final String QUALITY_UNSPECIFIED = "unspecified";

    private static final Gson GSON = new Gson();

    private final ConfigManager configManager;
    private final String instanceId;
    private final Uns uns;
    private final MessagingClient messaging;
    private final StreamSink streamSink; // nullable -> stream route falls back to local
    private final Clock clock;
    private volatile boolean warnedNoStream = false;

    /**
     * Library-internal constructor (see class javadoc). The identity-stamping and topic-minting
     * seams are pre-bound to one instance token.
     *
     * @param configManager the component config manager (envelope identity + {@code publish.channel})
     * @param instanceId    the instance token this facade is bound to
     * @param uns           the instance-bound UNS topic builder
     * @param messaging     the (guarded) messaging client
     * @param streamSink    the stream seam, or {@code null} when streaming is not configured
     * @param clock         the clock for {@code serverTs} defaults (injected for deterministic tests)
     */
    public DataFacade(ConfigManager configManager, String instanceId, Uns uns,
                      MessagingClient messaging, StreamSink streamSink, Clock clock) {
        this.configManager = Objects.requireNonNull(configManager, "configManager must not be null");
        this.instanceId = Objects.requireNonNull(instanceId, "instanceId must not be null");
        this.uns = Objects.requireNonNull(uns, "uns must not be null");
        this.messaging = Objects.requireNonNull(messaging, "messaging must not be null");
        this.streamSink = streamSink;
        this.clock = Objects.requireNonNull(clock, "clock must not be null");
    }

    // ===================== fluent builder entry points =====================

    /**
     * Starts building a {@code SouthboundSignalUpdate} for a stable {@code signal.id} — the fluent
     * body builder that subsumes the hand-assembled {@code JsonObject}. Terminate with
     * {@link SignalUpdate.Builder#publish()}.
     *
     * @param id the stable {@code signal.id} (REQUIRED — the consumer key)
     * @return the bound builder
     */
    public SignalUpdate.Builder signal(String id) {
        return new SignalUpdate.Builder(this, id);
    }

    // ===================== value shorthand =====================

    /**
     * The value-shorthand: publish one value for a signal path (the path doubles as the stable
     * {@code signal.id}). The single value is wrapped into a one-element {@code samples} array with
     * {@code quality=GOOD}, {@code qualityRaw="unspecified"}, {@code serverTs=now} — a caller never
     * emits a bare value.
     *
     * @param signalPath the signal path / stable id (e.g. {@code "press12/temperature"})
     * @param value      the measured value (REQUIRED)
     */
    public void publish(String signalPath, Object value) {
        signal(signalPath).addSample(value).signalPath(signalPath).publish();
    }

    /**
     * The value-shorthand with an explicit quality (so a source that knows the read is stale/failed
     * marks it {@code BAD}/{@code UNCERTAIN}).
     *
     * @param signalPath the signal path / stable id
     * @param value      the measured value (REQUIRED)
     * @param quality    the normalized quality
     */
    public void publish(String signalPath, Object value, Quality quality) {
        signal(signalPath).addSample(value, quality).signalPath(signalPath).publish();
    }

    // ===================== the raw escape hatch =====================

    /**
     * The raw escape hatch (D5): publishes a caller-owned pre-built body verbatim to
     * {@code data/{signalPath}}, applying <b>no</b> body defaulting — only the topic + identity
     * guarantees. For a component with an exotic body the facade should not shape.
     *
     * @param signalPath the signal path (sanitized into the channel)
     * @param body       the pre-built body, published untouched
     */
    public void publishBody(String signalPath, JsonObject body) {
        publishBody(signalPath, body, null);
    }

    /**
     * {@link #publishBody(String, JsonObject)} with an explicit {@link Channel} override.
     *
     * @param signalPath the signal path (sanitized into the channel)
     * @param body       the pre-built body, published untouched
     * @param via        the channel override, or {@code null} to resolve config ▸ LOCAL
     */
    public void publishBody(String signalPath, JsonObject body, Channel via) {
        Objects.requireNonNull(body, "body must not be null");
        String channel = channelToken(signalPath);
        String topic = uns.topic(UnsClass.DATA, channel);
        Message msg = message(body);
        route(via, topic, msg, signalPath, firstServerTsMillis(body));
    }

    // ===================== the SignalUpdate publish path =====================

    /**
     * Publishes a built {@link SignalUpdate}: validates {@code signal.id}, constructs the body with
     * the defaulting rules, sanitizes the path into the {@code data} channel, stamps the envelope,
     * and routes to the resolved channel.
     *
     * @param update the signal update
     * @throws IllegalArgumentException when {@code signal.id} is missing/empty or a sample carries
     *                                  no value, or the signal path is empty
     */
    public void publish(SignalUpdate update) {
        Objects.requireNonNull(update, "update must not be null");
        if (update.signalId() == null || update.signalId().isEmpty()) {
            throw new IllegalArgumentException("data publish requires a stable signal.id"
                    + " (the consumer key) - it is the only non-defaultable field");
        }
        if (update.samples().isEmpty()) {
            throw new IllegalArgumentException("data publish requires at least one sample");
        }
        JsonObject body = buildBody(update);
        String channel = channelToken(update.effectiveSignalPath());
        String topic = uns.topic(UnsClass.DATA, channel);
        Message msg = message(body);
        route(update.via(), topic, msg, update.signalId(), firstServerTsMillis(body));
    }

    // ===================== body construction (THE contract) =====================

    /**
     * Constructs the wire body from a {@link SignalUpdate}, applying the §2.1 defaulting rules
     * (quality → {@code GOOD} + {@code qualityRaw} marker, {@code serverTs} → now, samples wrapper).
     * Deterministic given the injected clock — this is the exact body the vectors pin.
     *
     * @param update the signal update (its {@code signal.id} must be set)
     * @return the constructed {@code SouthboundSignalUpdate} body
     * @throws IllegalArgumentException when a sample carries no value
     */
    public JsonObject buildBody(SignalUpdate update) {
        JsonObject signal = new JsonObject();
        signal.addProperty("id", update.signalId());
        if (update.signalName() != null) {
            signal.addProperty("name", update.signalName());
        }
        if (update.signalAddress() != null) {
            signal.add("address", update.signalAddress());
        }

        JsonArray samples = new JsonArray();
        for (SignalUpdate.Sample sample : update.samples()) {
            samples.add(buildSample(sample));
        }

        JsonObject body = new JsonObject();
        if (update.device() != null) {
            body.add("device", update.device());
        }
        body.add("signal", signal);
        body.add("samples", samples);
        return body;
    }

    /** Builds one sample with the quality/qualityRaw/serverTs defaulting rules. */
    private JsonObject buildSample(SignalUpdate.Sample sample) {
        if (sample.value() == null) {
            throw new IllegalArgumentException("data sample value is required (a quality-only"
                    + " sample is not a sample) - pass BAD/UNCERTAIN for a failed read");
        }
        JsonObject out = new JsonObject();
        out.add("value", toJsonElement(sample.value()));

        boolean qualityDefaulted = sample.quality() == null;
        Quality quality = qualityDefaulted ? Quality.GOOD : sample.quality();
        out.addProperty("quality", quality.wire());

        String qualityRaw = sample.qualityRaw();
        if (qualityRaw == null && qualityDefaulted) {
            qualityRaw = QUALITY_UNSPECIFIED;
        }
        if (qualityRaw != null) {
            out.addProperty("qualityRaw", qualityRaw);
        }

        if (sample.sourceTs() != null) {
            out.addProperty("sourceTs", sample.sourceTs());
        }
        out.addProperty("serverTs", sample.serverTs() != null ? sample.serverTs() : nowIso());
        return out;
    }

    // ===================== channel routing =====================

    /**
     * Resolves the effective channel: per-call {@code via} override ▸ config {@code publish.channel}
     * (instance ▸ global) ▸ {@link Channel#LOCAL} (DESIGN-class-facades §4, D1).
     *
     * @param via the per-call override, or {@code null}
     * @return the resolved channel (never {@code null})
     */
    public Channel resolveChannel(Channel via) {
        if (via != null) {
            return via;
        }
        Channel configured = configuredChannel();
        return configured != null ? configured : Channel.LOCAL;
    }

    /**
     * Reads the config {@code publish.channel} default (Option C): the bound instance's
     * {@code publish.channel} ▸ the global {@code component.global.publish.channel}. Best-effort —
     * any lookup/parse anomaly yields {@code null} (fall through to LOCAL).
     */
    private Channel configuredChannel() {
        try {
            Channel fromInstance = publishChannelOf(configManager.getInstanceConfig(instanceId));
            if (fromInstance != null) {
                return fromInstance;
            }
            return publishChannelOf(configManager.getGlobalConfig());
        } catch (RuntimeException e) {
            LOGGER.debug("publish.channel lookup failed (defaulting to LOCAL): {}", e.toString());
            return null;
        }
    }

    /** {@code section.publish.channel} as a {@link Channel}, or {@code null} when absent/unparseable. */
    private static Channel publishChannelOf(JsonObject section) {
        if (section == null || !section.has("publish") || !section.get("publish").isJsonObject()) {
            return null;
        }
        JsonObject publish = section.getAsJsonObject("publish");
        if (!publish.has("channel") || !publish.get("channel").isJsonPrimitive()) {
            return null;
        }
        return Channel.fromConfig(publish.get("channel").getAsString());
    }

    /**
     * Routes a built envelope to the resolved channel. LOCAL publishes on the guarded bus;
     * NORTHBOUND publishes to IoT Core; a stream route appends the serialized envelope to the named
     * stream (falling back to LOCAL when no sink is wired). Northbound / stream failures are
     * caught + logged (they must never flip local readiness).
     */
    private void route(Channel via, String topic, Message msg, String partitionKey, long tsMillis) {
        Channel channel = resolveChannel(via);
        switch (channel.kind()) {
            case LOCAL -> messaging.publish(topic, msg);
            case NORTHBOUND -> {
                try {
                    messaging.publishNorthbound(topic, msg, Qos.AT_LEAST_ONCE);
                } catch (Exception e) {
                    LOGGER.warn("Northbound data publish on '{}' failed (local readiness"
                            + " unaffected): {}", topic, e.toString());
                }
            }
            case STREAM -> appendToStream(channel.streamName(), topic, msg, partitionKey, tsMillis);
        }
    }

    /** The {@code stream:<name>} route: append the serialized envelope, or fall back to LOCAL. */
    private void appendToStream(String streamName, String topic, Message msg, String partitionKey,
                                long tsMillis) {
        if (streamSink == null) {
            if (!warnedNoStream) {
                warnedNoStream = true;
                LOGGER.warn("data channel 'stream:{}' requested but streaming is not configured -"
                        + " routing to LOCAL instead (readiness/no-streaming -> local)", streamName);
            }
            messaging.publish(topic, msg);
            return;
        }
        try {
            byte[] payload = msg.toDict().toString().getBytes(StandardCharsets.UTF_8);
            streamSink.append(streamName, partitionKey, tsMillis, payload);
        } catch (Exception e) {
            LOGGER.warn("Stream append to 'stream:{}' failed (local readiness unaffected): {}",
                    streamName, e.toString());
        }
    }

    // ===================== helpers =====================

    /** The sanitized channel token for a signal path (each {@code /}-token → a UNS token). */
    public String channelToken(String signalPath) {
        if (signalPath == null || signalPath.isEmpty()) {
            throw new IllegalArgumentException("data signal path must be non-empty");
        }
        String[] tokens = signalPath.split("/", -1);
        StringBuilder sb = new StringBuilder(signalPath.length());
        for (int i = 0; i < tokens.length; i++) {
            if (i > 0) {
                sb.append('/');
            }
            sb.append(ConfigManager.sanitize(tokens[i]));
        }
        return sb.toString();
    }

    /** Builds the identity-stamped envelope with the signal-update header. */
    private Message message(JsonObject body) {
        return MessageBuilder.create(DATA_MESSAGE_NAME, DATA_MESSAGE_VERSION)
                .withConfig(configManager)
                .withInstance(instanceId)
                .withPayload(body)
                .build();
    }

    /** ISO-8601 UTC ({@code …Z}) "now" from the injected clock. */
    private String nowIso() {
        return Instant.now(clock).toString();
    }

    /** The first sample's {@code serverTs} as epoch millis (the stream record timestamp). */
    private static long firstServerTsMillis(JsonObject body) {
        try {
            JsonArray samples = body.getAsJsonArray("samples");
            if (samples != null && !samples.isEmpty()) {
                JsonObject first = samples.get(0).getAsJsonObject();
                if (first.has("serverTs")) {
                    return Instant.parse(first.get("serverTs").getAsString()).toEpochMilli();
                }
            }
        } catch (RuntimeException ignored) {
            // Fall through to now() below.
        }
        return System.currentTimeMillis();
    }

    /** A JSON-native value: a {@link JsonElement} verbatim, else via Gson's reflective adapter. */
    private static JsonElement toJsonElement(Object value) {
        if (value instanceof JsonElement element) {
            return element;
        }
        return GSON.toJsonTree(value);
    }

    /** The instance token this facade is bound to. */
    public String instanceId() {
        return instanceId;
    }
}

File: libs/ts/src/facades/data_facade.ts
/**
 * The `data()` publish facade — the telemetry / signal data plane (DESIGN-class-facades §2.1,
 * D2/D5). It **constructs and validates the `SouthboundSignalUpdate` body**
 * (`device`/`signal`/`samples`) so an adapter never hand-builds it, applies the body defaults,
 * sanitizes the signal path into the UNS `data` channel, stamps the envelope identity, and
 * routes to the resolved {@link Channel}. It publishes through the **ordinary, guarded**
 * `messaging().publish(...)` — `data` is non-reserved, so it passes the guard; the facade adds
 * body-contract enforcement + defaults, **not** privilege (there is no reserved-publish seam
 * here). Mirrors the Java `com.mbreissi.edgecommons.facades.DataFacade`.
 *
 * Body (`header.name` = {@link DATA_MESSAGE_NAME}, version {@link DATA_MESSAGE_VERSION}):
 * ```jsonc
 * { "device": {adapter, instance, endpoint}?,
 *   "signal": { "id": "<REQUIRED>", "name"?, "address"? },
 *   "samples": [ { "value": "<REQUIRED>", "quality", "qualityRaw"?, "sourceTs"?, "serverTs" } ] }
 * ```
 *
 * Defaulting (DESIGN-class-facades §2.1, pinned by `uns-test-vectors/data.json`):
 * 1. `quality` → {@link Quality.Good} when omitted on a sample that carries a value.
 * 2. `qualityRaw` → the synthetic marker {@link QUALITY_UNSPECIFIED} when (and only when) the
 *    quality was defaulted; else the caller's value verbatim, else absent.
 * 3. `serverTs` → now (ISO-8601 UTC `…Z`, from the injected clock) when omitted; `sourceTs` is
 *    **never** synthesized (absent when the source has none).
 * 4. The `samples` wrapper is enforced for the value-shorthand (a caller never emits a bare value).
 * 5. `signal.id` is the **only** hard reject — a publish with no stable id throws
 *    `EdgeCommonsError` (kind `Validation`) at the call site.
 *
 * Channel routing (DESIGN-class-facades §4, D1): per-call {@link SignalUpdate.via} override ▸
 * config `publish.channel` (instance ▸ global) ▸ {@link Channel.LOCAL}. A `stream:<name>` route
 * serializes the same envelope and appends it via the bound {@link StreamSink} (partition key =
 * `signal.id`, ts = `serverTs`); when streaming is not configured it falls back to a LOCAL
 * publish (readiness / no-streaming → local). Northbound / stream transport failures are caught
 * and logged — they must never flip local readiness.
 *
 * **Library-internal:** obtain the bound instance via `gg.instance(id).data()` (or the
 * `main`-instance convenience `gg.data()`); the exported constructor exists only so
 * `EdgeCommonsInstance`/`EdgeCommons` (`edgecommons.ts`) can wire it.
 */
import type { Config } from "../config/model";
import { sanitize } from "../config/template";
import { EdgeCommonsError } from "../errors";
import { logger } from "../logging";
import type { Message } from "../message";
import { MessageBuilder } from "../message";
import type { IMessagingService } from "../messaging/types";
import { Qos } from "../messaging/types";
import { Uns, UnsClass } from "../uns";
import { Channel } from "./channel";
import { type ClockMillis, toIso } from "./clock";
import { Quality } from "./quality";
import type { Sample, SignalUpdate } from "./signal_update";
import { SignalUpdateBuilder } from "./signal_update";
import type { StreamSink } from "./stream_sink";

/** The signal-update envelope header name (`docs/SOUTHBOUND.md` §2). */
export const DATA_MESSAGE_NAME = "SouthboundSignalUpdate";
/** The signal-update envelope header version. */
export const DATA_MESSAGE_VERSION = "1.0";
/** The `qualityRaw` marker written when `quality` was defaulted to {@link Quality.Good}. */
export const QUALITY_UNSPECIFIED = "unspecified";

export class DataFacade {
  private warnedNoStream = false;

  /**
   * @param configProvider a snapshot accessor (envelope identity + `publish.channel` lookup)
   * @param instanceId     the instance token this facade is bound to
   * @param uns            the instance-bound UNS topic builder
   * @param messaging      the (guarded) messaging service
   * @param streamSink     the stream seam, or `undefined` when streaming is not configured
   * @param clockMillis    the clock for `serverTs` defaults (injected for deterministic tests)
   */
  constructor(
    private readonly configProvider: () => Config,
    private readonly instanceId: string,
    private readonly uns: Uns,
    private readonly messaging: IMessagingService,
    private readonly streamSink: StreamSink | undefined,
    private readonly clockMillis: ClockMillis = () => Date.now(),
  ) {}

  /** The instance token this facade is bound to. */
  instanceIdValue(): string {
    return this.instanceId;
  }

  // ===================== fluent builder entry point =====================

  /**
   * Starts building a `SouthboundSignalUpdate` for a stable `signal.id` — the fluent body
   * builder that subsumes the hand-assembled JSON object. Terminate with
   * `SignalUpdateBuilder.publish()`.
   *
   * @param id the stable `signal.id` (REQUIRED at publish — the consumer key)
   */
  signal(id: string | undefined): SignalUpdateBuilder {
    return new SignalUpdateBuilder(id, (update) => this.publishUpdate(update));
  }

  // ===================== publish (shorthand + full form) =====================

  /**
   * The value-shorthand: publish one value for a signal path (the path doubles as the stable
   * `signal.id`). The single value is wrapped into a one-element `samples` array with
   * `quality=GOOD`, `qualityRaw="unspecified"`, `serverTs=now` — a caller never emits a bare value.
   */
  async publish(signalPath: string, value: unknown, quality?: Quality): Promise<void>;
  /** Publishes a built {@link SignalUpdate} (the full form; see {@link signal}/{@link SignalUpdateBuilder.build}). */
  async publish(update: SignalUpdate): Promise<void>;
  async publish(a: string | SignalUpdate, value?: unknown, quality?: Quality): Promise<void> {
    if (typeof a === "string") {
      const update = this.signal(a).addSample(value, { quality }).signalPath(a).build();
      await this.publishUpdate(update);
    } else {
      await this.publishUpdate(a);
    }
  }

  // ===================== the raw escape hatch =====================

  /**
   * The raw escape hatch (D5): publishes a caller-owned pre-built body verbatim to
   * `data/{signalPath}`, applying **no** body defaulting — only the topic + identity guarantees.
   * For a component with an exotic body the facade should not shape.
   */
  async publishBody(signalPath: string, body: Record<string, unknown>, via?: Channel): Promise<void> {
    const channel = this.channelToken(signalPath);
    const topic = this.uns.topic(UnsClass.Data, channel);
    const msg = this.message(body);
    await this.route(via, topic, msg, signalPath, firstServerTsMillis(body));
  }

  // ===================== the SignalUpdate publish path =====================

  /**
   * Validates `signal.id`, constructs the body with the defaulting rules, sanitizes the path
   * into the `data` channel, stamps the envelope, and routes to the resolved channel.
   *
   * @throws EdgeCommonsError (kind `Validation`) when `signal.id`/the path is missing/empty, there are no
   *                 samples, or a sample carries no value
   */
  private async publishUpdate(update: SignalUpdate): Promise<void> {
    if (!update.signalId) {
      throw EdgeCommonsError.validation(
        "data publish requires a stable signal.id (the consumer key) - it is the only" +
          " non-defaultable field",
      );
    }
    if (update.samples.length === 0) {
      throw EdgeCommonsError.validation("data publish requires at least one sample");
    }
    const body = this.buildBody(update);
    const channel = this.channelToken(update.signalPath ?? update.signalId);
    const topic = this.uns.topic(UnsClass.Data, channel);
    const msg = this.message(body);
    await this.route(update.via, topic, msg, update.signalId, firstServerTsMillis(body));
  }

  // ===================== body construction (THE contract) =====================

  /**
   * Constructs the wire body from a {@link SignalUpdate}, applying the §2.1 defaulting rules
   * (quality → GOOD + `qualityRaw` marker, `serverTs` → now, samples wrapper). Deterministic
   * given the injected clock — this is the exact body the vectors pin.
   *
   * @throws EdgeCommonsError (kind `Validation`) when a sample carries no value
   */
  buildBody(update: SignalUpdate): Record<string, unknown> {
    const signal: Record<string, unknown> = { id: update.signalId };
    if (update.signalName !== undefined) signal.name = update.signalName;
    if (update.signalAddress !== undefined) signal.address = update.signalAddress;

    const samples = update.samples.map((sample) => this.buildSample(sample));

    const body: Record<string, unknown> = {};
    if (update.device !== undefined) body.device = update.device;
    body.signal = signal;
    body.samples = samples;
    return body;
  }

  /** Builds one sample with the quality/qualityRaw/serverTs defaulting rules. */
  private buildSample(sample: Sample): Record<string, unknown> {
    if (sample.value === undefined) {
      throw EdgeCommonsError.validation(
        "data sample value is required (a quality-only sample is not a sample) - pass" +
          " BAD/UNCERTAIN for a failed read",
      );
    }
    const out: Record<string, unknown> = { value: sample.value };

    const qualityDefaulted = sample.quality === undefined;
    const quality = qualityDefaulted ? Quality.Good : sample.quality;
    out.quality = quality;

    let qualityRaw = sample.qualityRaw;
    if (qualityRaw === undefined && qualityDefaulted) {
      qualityRaw = QUALITY_UNSPECIFIED;
    }
    if (qualityRaw !== undefined) out.qualityRaw = qualityRaw;

    if (sample.sourceTs !== undefined) out.sourceTs = sample.sourceTs;
    out.serverTs = sample.serverTs ?? toIso(this.clockMillis());
    return out;
  }

  // ===================== channel routing =====================

  /**
   * Resolves the effective channel: per-call `via` override ▸ config `publish.channel`
   * (instance ▸ global) ▸ {@link Channel.LOCAL} (DESIGN-class-facades §4, D1).
   */
  resolveChannel(via: Channel | undefined): Channel {
    if (via !== undefined) return via;
    return this.configuredChannel() ?? Channel.LOCAL;
  }

  /**
   * Reads the config `publish.channel` default (Option C): the bound instance's
   * `publish.channel` ▸ the global `component.global.publish.channel`. Best-effort — any
   * lookup/parse anomaly yields `undefined` (fall through to LOCAL).
   */
  private configuredChannel(): Channel | undefined {
    try {
      const config = this.configProvider();
      const fromInstance = publishChannelOf(config.instance(this.instanceId));
      if (fromInstance !== undefined) return fromInstance;
      return publishChannelOf(config.global());
    } catch (e) {
      logger.debug(`publish.channel lookup failed (defaulting to LOCAL): ${errMsg(e)}`);
      return undefined;
    }
  }

  /**
   * Routes a built envelope to the resolved channel. LOCAL publishes on the guarded bus;
   * NORTHBOUND publishes to IoT Core; a stream route appends the serialized envelope to the
   * named stream (falling back to LOCAL when no sink is wired). Northbound / stream failures are
   * caught + logged (they must never flip local readiness).
   */
  private async route(
    via: Channel | undefined,
    topic: string,
    msg: Message,
    partitionKey: string,
    tsMillis: number,
  ): Promise<void> {
    const channel = this.resolveChannel(via);
    switch (channel.kind) {
      case "local":
        await this.messaging.publish(topic, msg);
        break;
      case "northbound":
        try {
          await this.messaging.publishNorthbound(topic, msg, Qos.AtLeastOnce);
        } catch (e) {
          logger.warn(`Northbound data publish on '${topic}' failed (local readiness unaffected): ${errMsg(e)}`);
        }
        break;
      case "stream":
        await this.appendToStream(channel.streamName, topic, msg, partitionKey, tsMillis);
        break;
    }
  }

  /** The `stream:<name>` route: append the serialized envelope, or fall back to LOCAL. */
  private async appendToStream(
    streamName: string,
    topic: string,
    msg: Message,
    partitionKey: string,
    tsMillis: number,
  ): Promise<void> {
    if (!this.streamSink) {
      if (!this.warnedNoStream) {
        this.warnedNoStream = true;
        logger.warn(
          `data channel 'stream:${streamName}' requested but streaming is not configured -` +
            " routing to LOCAL instead (readiness/no-streaming -> local)",
        );
      }
      await this.messaging.publish(topic, msg);
      return;
    }
    try {
      const payload = msg.toBytes();
      this.streamSink(streamName, partitionKey, tsMillis, payload);
    } catch (e) {
      logger.warn(`Stream append to 'stream:${streamName}' failed (local readiness unaffected): ${errMsg(e)}`);
    }
  }

  // ===================== helpers =====================

  /** The sanitized channel token for a signal path (each `/`-token → a UNS token). */
  channelToken(signalPath: string | undefined): string {
    if (!signalPath) {
      throw EdgeCommonsError.validation("data signal path must be non-empty");
    }
    return signalPath
      .split("/")
      .map((token) => sanitize(token))
      .join("/");
  }

  /** Builds the identity-stamped envelope with the signal-update header. */
  private message(body: unknown): Message {
    return MessageBuilder.create(DATA_MESSAGE_NAME, DATA_MESSAGE_VERSION)
      .withConfig(this.configProvider())
      .withInstance(this.instanceId)
      .withSouthboundSignalUpdate(body)
      .build();
  }
}

/** `section.publish.channel` as a {@link Channel}, or `undefined` when absent/unparseable. */
function publishChannelOf(section: unknown): Channel | undefined {
  if (section === null || typeof section !== "object") return undefined;
  const publish = (section as Record<string, unknown>).publish;
  if (publish === null || typeof publish !== "object") return undefined;
  const channel = (publish as Record<string, unknown>).channel;
  return typeof channel === "string" ? Channel.fromConfig(channel) : undefined;
}

/** The first sample's `serverTs` as epoch millis (the stream record timestamp). */
function firstServerTsMillis(body: Record<string, unknown>): number {
  const samples = body.samples;
  if (Array.isArray(samples) && samples.length > 0) {
    const first = samples[0] as Record<string, unknown>;
    if (typeof first.serverTs === "string") {
      const parsed = Date.parse(first.serverTs);
      if (!Number.isNaN(parsed)) return parsed;
    }
  }
  return Date.now();
}

function errMsg(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}
